clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchMetricsManagerTest.java

```diff
@@ -145,46 +167,209 @@ public void testRecordsFetched() {
+        TopicPartition tp3 = new TopicPartition("another.topic", 1);
+        int initialMetricsSize = metrics.metrics().size();
+
+        SubscriptionState subscriptionState = new SubscriptionState(new LogContext(), AutoOffsetResetStrategy.NONE);
+        subscriptionState.assignFromUser(Set.of(tp1));
+
+        metricsManager.maybeUpdateAssignment(subscriptionState);
+        // 1 new metrics shall be registered.
+        assertEquals(1, metrics.metrics().size() - initialMetricsSize);
+
+        subscriptionState.assignFromUser(Set.of(tp1, tp2));
+        subscriptionState.updatePreferredReadReplica(tp2, 1, () -> 0L);
+        metricsManager.maybeUpdateAssignment(subscriptionState);
+        // Another 2 metrics gets registered as deprecated metrics should be reported for tp2.
+        assertEquals(3, metrics.metrics().size() - initialMetricsSize);
+
+        Map<String, String> tags1 = Map.of("topic", tp1.topic(), "partition", String.valueOf(tp1.partition()));
+        Map<String, String> tags2 = Map.of("topic", tp2.topic(), "partition", String.valueOf(tp2.partition()));
+        Map<String, String> deprecatedTags = topicPartitionTags(tp2);
+        // Validate preferred read replica metrics.
+        assertEquals(-1, readReplicaMetricValue(metricsRegistry.partitionPreferredReadReplica, tags1), EPSILON);
+        assertEquals(1, readReplicaMetricValue(metricsRegistry.partitionPreferredReadReplica, tags2), EPSILON);
+        assertEquals(1, readReplicaMetricValue(metricsRegistry.partitionPreferredReadReplica, deprecatedTags), EPSILON);
+
+        // Remove tp2 from subscription set.
+        subscriptionState.assignFromUser(Set.of(tp1, tp3));
+        metricsManager.maybeUpdateAssignment(subscriptionState);
+        // Metrics count shall remain same as tp2 should be removed and tp3 gets added.
+        assertEquals(3, metrics.metrics().size() - initialMetricsSize);
+
+        // Remove all partitions.
+        subscriptionState.assignFromUser(Set.of());
+        metricsManager.maybeUpdateAssignment(subscriptionState);
+        // Metrics count shall be same as initial count as all new metrics shall be removed.
+        assertEquals(initialMetricsSize, metrics.metrics().size());
+    }
+
+    @Test
+    public void testMaybeUpdateAssignmentWithAdditionalRegisteredMetrics() {
+        TopicPartition tp1 = new TopicPartition(TOPIC_NAME, 0);
+        TopicPartition tp2 = new TopicPartition("another.topic", 0);
+        TopicPartition tp3 = new TopicPartition("another.topic", 1);
+
+        int initialMetricsSize = metrics.metrics().size();
+
+        metricsManager.recordPartitionLag(tp1, 14);
+        metricsManager.recordPartitionLead(tp1, 11);
+        metricsManager.recordPartitionLag(tp2, 5);
+        metricsManager.recordPartitionLead(tp2, 1);
+        metricsManager.recordPartitionLag(tp3, 4);
+        metricsManager.recordPartitionLead(tp3, 2);
+
+        int additionalRegisteredMetricsSize = metrics.metrics().size();
```

Comment:
Wouldn't it be sensible to assert the number of additional registered metrics too?

Reply:
I haven't asserted them on purpose as above tests already validated them.